Add strict foldlWithKeyM

andrewthad opened the pull request:
Resolves https://github.com/haskell/containers/issues/793

This is just the implementation of a single monadic fold. @treeowl if you confirm that this implementation of this particular function looks good, I can add all of the lazy folds and the right folds as well

andrewthad commented:
@treeowl Do you have any feedback on this?

andrewthad commented:
I've changed this in the most recent commit. I think I'll have:

* left monadic fold on map
* right monadic fold on map
* left monadic fold on intmap
* right monadic fold on intmap

Anything else you'd like to see in that list.

treeowl commented on containers/src/Data/Map/Internal.hs:
```diff
@@ -3365,16 +3365,16 @@ foldlWithKey' f z = go z
-    go !z' Tip             = return z'
+    go z' Tip             = return z'
     go z' (Bin _ kx x l r) = do
       !z'' <- go z' l
```
This is still strict. I guess it would be okay to have both a lazy and strict version if you want, though it seems a bit redundant.

andrewthad replied:
Sorry, that was an oversight. This should be resolved in the latest commit.

treeowl commented:
Yeah, could you get me some pizza? I've been missing the really good stuff.

On Wed, Aug 25, 2021, 12:27 PM Andrew Martin ***@***.***>
wrote:

> I've changed this in the most recent commit. I think I'll have:
>
>    - left monadic fold on map
>    - right monadic fold on map
>    - left monadic fold on intmap
>    - right monadic fold on intmap
>
> Anything else you'd like to see in that list.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/containers/pull/795#issuecomment-905688155>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOOF7NXUFRWXYXLSDOGJFLT6UKYZANCNFSM5CH5OETQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

andrewthad commented:
I live in Atlanta, so all the truly good pizza is a ways up the east coast from me.

treeowl commented:
Oh darn. Guess I'll have to make you pizza if you come visit (D.C. area).

On Wed, Aug 25, 2021, 12:33 PM Andrew Martin ***@***.***>
wrote:

> I live in Atlanta, so all the truly good pizza is a ways up the east coast
> from me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/containers/pull/795#issuecomment-905692275>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOOF7IDZIB4557ALSMBIOLT6ULPFANCNFSM5CH5OETQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

ocramz commented:
@treeowl This seems like an uncontroversial (and welcome) addition, how can we move this PR forward?

andrewthad commented:
Is there anything I can do to move this forward?

treeowl commented:
I got sick a few weeks ago, and then went on family vacation. I expect to be ready to get back to my GitHub backlog by next Thursday. If this is ready for review, could you mark it as such?

andrewthad commented:
I've squashed and rebased. This adds `foldlWithKeyM` and `foldrWithKeyM` functions for the `Data.Map` data structure, both of which are lazy in the accumulator. It's ready for review.
